bin/qdbg-vsc-adapter

```diff
@@ -0,0 +1,385 @@
+#!/usr/bin/env qore
+# -*- mode: qore; indent-tabs-mode: nil -*-
+
+%enable-all-warnings
+%new-style
+
+%requires qore >= 0.8.13
+%requires VscDebugAdapter
+%requires DebugProgramControl >= 0.1.2
+
+%requires WebSocketClient
+%requires yaml >= 0.5
+%requires ConnectionProvider
+
+%allow-debugger
+%no-debugging
+
+%exec-class DebugWrapper
+
+sub get_full_log_file_path(string ext) {
+    if (PlatformOS == "Windows") {
+        return getenv("APPDATA") + DirSep + "QoreDebugAdapter" + DirSep + "qda" + ext;
+    } else {
+        return getenv("HOME") + DirSep + ".qoredebugadapter" + ext;
+    }
+}
+
+sub printErr(string fmt) {
+    string s = vsprintf(fmt, argv)+"\r\n";
+    stderr.print(s);
+%ifdef VDA_EARLY_LOG
+    # in case of an exception when starting adapter the VSC does not print stderr output
+    # so get what is wrong is pain, intended to be uncommented only when required
+    # Infamous "Debug adapter process has terminated unexpectedly" is such a case
+    File f();
+    f.open(get_full_log_file_path(".stderr"), O_WRONLY|O_APPEND|O_CREAT);  # no O_TRUNC
+    f.printf("[pid %d] %s", getpid(), s);
+    f.close();
+%endif
+}
+
+class VscDebugAdapterLocal inherits VscDebugAdapter {
+    public {
+        DebugProgramControlLocal dpc;
+        # increment when command is sent and decrement when response is received
+        Counter requestFlag();
+        any threadData;
+        *hash programOpts;
+    }
+    constructor(DebugProgramControlLocal n_dpc, hash opts): VscDebugAdapter() {
+        dpc = n_dpc;
+        dpc.vda = self;
```

Comment:
there is a circular reference between objects, why?

Reply:
Sure, because they need access data from opposite instance.